students/Kristy_Martini/lesson09/mailroom.py

```diff
@@ -0,0 +1,149 @@
+        for donor in new_report.sorted_list:
+            new_report.sorted_dict[donor.name] = donor
+
+    def create_report(self, new_report=None):
+        """ Display donor report to the user"""
+        if new_report is None:
+            new_report = Report()
+        new_report.sort_donors(new_report)
+        report_lines = []
+        report_lines.append("Donor Name          | Total Given   | Num Gifts | Average Gift")
+        report_lines.append("--------------------------------------------------------------")
+        for value in new_report.sorted_dict.values():
+            new_line = '{0:21}'.format(value.name) + "$" + '{0:14}'.format(value.total_gift_value) + '{0:12}'.format(value.num_gifts) + " $" + '{0:12}'.format(value.average_gift)
+            report_lines.append(new_line)
+        report_lines.append("\n")
+        return report_lines
+
+    def display_report(self, new_report):
+        report_lines = new_report.create_report(new_report)
+        for line in report_lines:
+            terminal_print(line)
+        prompt_user(new_report)
+
+class Donor:
+    """ Donor class stores the info associated with each donor"""
+    def __init__(self, name, gift_value=0, num_gifts=1, total_gift_value=0):
+        self.name = name
+        self.gift_value = gift_value
+        self.num_gifts = num_gifts
+        self.total_gift_value = total_gift_value
+        try:
+            self.average_gift = self.total_gift_value/self.num_gifts
+        except ZeroDivisionError:
+            self.average_gift = 0
+
+    def add_gift(self, amount):
+        """ Add gift amount to a donor already present in database"""
+        self.total_gift_value = self.total_gift_value + amount
+        self.num_gifts += 1
+        self.average_gift = self.total_gift_value/self.num_gifts
+
+def check_name(new_report=None):
+    """ Check the name of the donor before sending a thank you, add to donor list if they do not exist"""
+    if new_report is None:
+        new_report = Report()
+    terminal_print("To whom would you like to send a thank you?")
+    name_to_thank = input_call("Please enter the full name of the donor you'd like to thank, or enter 'list' to view a list of current donors: ")
+    if name_to_thank == "quit":
+        prompt_user(new_report)
+    if name_to_thank == "list":
+        new_report.display_report(new_report)
+        terminal_print("Would you like to thank a donor from this list?")
+        name_to_thank= input_call("Please enter the full name of the donor you'd like to thank, or enter 'quit' to exit: ")
+        if name_to_thank == "quit":
+            prompt_user(new_report)
+    try: 
+        donor = new_report.donors[name_to_thank]
+    except KeyError:
+        terminal_print(name_to_thank, " is a new donor. They will be added to our database.")
+        newDonor = Donor(name_to_thank, 0, 0, 0)
+        new_report.add_donor(newDonor)
+        donor = new_report.donors[newDonor.name]
```

Comment:
you should have a method for this access -- remember that we don't want users to have this implementation detail if we decide to change the data structure.